firebase-firestore/src/main/java/com/google/firebase/firestore/util/ListenerRegistrationImpl.java

```diff
@@ -33,6 +35,115 @@
   /** The event listener for the query that raises events asynchronously. */
   private final ExecutorEventListener<ViewSnapshot> asyncEventListener;
 
+  public static class StopListenerSupportFragment extends android.support.v4.app.Fragment {
+    @Override
+    public void onStop() {
+      super.onStop();
+      for (Runnable callback : callbacks) {
+        if (callback != null) {
+          callback.run();
+        }
+      }
+    }
+
+    void addCallback(Runnable callback) {
+      callbacks.add(callback);
+    }
+
+    private List<Runnable> callbacks = new ArrayList();
+  }
+
+  public static class StopListenerFragment extends android.app.Fragment {
+    @Override
+    public void onStop() {
+      super.onStop();
+      for (Runnable callback : callbacks) {
+        if (callback != null) {
+          callback.run();
+        }
+      }
+    }
+
+    void addCallback(Runnable callback) {
+      callbacks.add(callback);
+    }
+
+    private List<Runnable> callbacks = new ArrayList();
+  }
+
+  private static final String SUPPORT_FRAGMENT_TAG = "FirestoreOnStopObserverSupportFragment";
+  private static final String FRAGMENT_TAG = "FirestoreOnStopObserverFragment";
+
+  private void onActivityStopCallOnce(Activity activity, Runnable callback) {
```

Comment:
This is a massive method and I think there would be benefits to carving it up into two and picking which one to call in the ListenerRegistrationImpl  constructor. There it can figure out which kind of activity we're working with, cast, and then delegate to the the specific implementations (which would remain essentially replicas of each other).

The benefits are:
  * Within the implementation only the specific type will be in scope (i.e. the FragmentActivity variant will take that as a parameter)
  * Within the implementation that uses deprecated methods we can suppress deprecation warnings but have those only affect that code (we'll still see deprecation warnings if FragmentActivity stuff is deprecated).